x/valset-pref/msg_server_test.go

```diff
@@ -0,0 +1,110 @@
+package keeper_test
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	valPref "github.com/osmosis-labs/osmosis/v12/x/valset-pref"
+	"github.com/osmosis-labs/osmosis/v12/x/valset-pref/types"
+)
+
+func (suite *KeeperTestSuite) TestSetValidatorSetPreference() {
+	suite.SetupTest()
+
+	// setup 3 validators
+	valAddrs := suite.SetupMultipleValidators(3)
+
+	tests := []struct {
+		name        string
+		delegator   sdk.AccAddress
+		preferences []types.ValidatorPreference
+		expectPass  bool
+	}{
+		{
+			name:      "creation of new validator set",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name:      "update existing validator with same valAddr and weights",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name:      "update existing validator with same valAddr but different weights",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+			},
+			expectPass: true,
```

Comment:
Would this pass if we kept valAddrs[0] at 5 and only updated [1] and [2]?

Reply:
I think I answered my question after reading validator_set.go. The answer to this is "no" but IMO we should make it so it is "yes".

Reply:
Yup covered this case here [`f729ac0`](https://github.com/osmosis-labs/osmosis/pull/2892/commits/f729ac0bb815359d167d5e1897815297624052dd)